File: keyboard/Sources/Prediction/SymSpell/SymSpellProvider.swift
import Foundation

/// Adapts SymSpell to the SuggestionProvider protocol.
///
/// Handles the fusion rule:
/// - If `lookupWord` is a real word (trie hit) → prefix completions from trie.
/// - If `lookupWord` is a typo (no trie hit) → SymSpell corrections.
///
/// Capitalization is inferred from `currentWord` and applied to suggestions
/// so that the input chip and suggestion chips appear consistent.
final class SymSpellProvider: SuggestionProvider {

    private let symSpell: SymSpell
    private let trie: Trie

    init(symSpell: SymSpell, trie: Trie) {
        self.symSpell = symSpell
        self.trie = trie
    }

    /// Applies the capitalization pattern from `input` to `suggestion`.
    ///
    /// Heuristic (covers ~95% of real cases):
    /// - If suggestion already contains an uppercase letter beyond position 0,
    ///   it is likely a proper noun (e.g. "USA", "iPhone") → return as-is.
    /// - If `input.first?.isUppercase == true` and the rest is lowercase →
    ///   "sentence case" → uppercase the first letter of suggestion.
    /// - If `input.allSatisfy(\.isUppercase)` and `input.count > 1` →
    ///   "all caps" → uppercase the entire suggestion.
    /// - Otherwise → "lowercase / mixed" → return suggestion as-is.
    static func applyCapitalizationTemplate(from input: String, to suggestion: String) -> String {
        // Preserve suggestions that are already capitalized (proper nouns / acronyms).
        let suggestionAfterFirst = suggestion.dropFirst()
        if suggestionAfterFirst.contains(where: { $0.isUppercase }) {
            return suggestion
        }

        guard let firstChar = input.first else { return suggestion }

        let rest = input.dropFirst()

        if firstChar.isUppercase, rest.allSatisfy({ $0.isLowercase || $0.isWhitespace }) {
            // Sentence case: capitalize first letter of suggestion.
            guard let sugFirst = suggestion.first else { return suggestion }
            return String(sugFirst.uppercased()) + suggestion.dropFirst()
        }

        if input.count > 1, input.allSatisfy(\.isUppercase) {
            // All caps: uppercase the entire suggestion.
            return suggestion.uppercased()
        }

        // Lowercase / mixed: leave suggestion as-is.
        return suggestion
    }

    func suggest(for context: SuggestionContext, limit: Int) -> [Suggestion] {
        let word = ApostropheNormalizer.canonicalize(context.lookupWord.lowercased())
        guard !word.isEmpty else { return [] }

        // Compute isRealWord once — drives both the verbatim-candidate display
        // signal and the prefix-completion vs typo-correction branch.
        // A word is "real" if it is in the built-in trie OR has been learned
        // by the user. Learned words must take the prefix-completion path so
        // SymSpell does not treat them as typos and outrank them with corrections.
        let isRealWord = trie.contains(word: word)
            || LearnedWordsStore.shared.contains(word)

        let contractionExpansion = Contractions.expansion(for: word)

        // When a contraction exists, it should be the PRIMARY candidate (leftmost
        // chip, highest score). The verbatim is demoted so it appears as a
        // secondary option the user can tap to keep the apostrophe-less form if
        // they really want it. The 0.5 score gap ensures the contraction survives
        // KenLM fusion (which blends at α=0.5, preserving a 0.25 gap after
        // blending — contraction always wins).
        let verbatimScore: Double = contractionExpansion != nil ? 0.5 : 1.0

        // Always include the input itself as a chip. Mark it as an unknown-verbatim
        // when the typed word is not a known/learned word — the UI renders these
        // with quotes so the user can distinguish the verbatim candidate from a
        // normal suggestion.
        var results: [Suggestion] = [
            Suggestion(
                text: context.currentWord,
                score: verbatimScore,
                source: .symspell,
                isUnknownVerbatim: !isRealWord
            )
        ]

        // Contraction fast-path: checked BEFORE trie/SymSpell so that
        // apostrophe-less forms like "dont" produce "don't" deterministically.
        // Real-word status is irrelevant here — "dont" IS a real word in our
        // lexicon but the user very likely meant "don't". Inserted at position 0
        // so it appears leftmost in the suggestion bar.
        if let contraction = contractionExpansion {
            let capped = Self.applyCapitalizationTemplate(from: context.currentWord, to: contraction)
            results.insert(
                Suggestion(
                    text: capped,
                    score: 1.0,
                    source: .contraction,
                    isUnknownVerbatim: false
                ),
                at: 0
            )
        }

        if isRealWord {
            // Prefix completions from trie.
            let completions = trie.suggest(prefix: word, limit: max(limit, 20))
            // Sort by frequency from SymSpell's canonical dictionary.
            let sorted = completions.sorted { (symSpell.dictionary[$0.lowercased()] ?? 0) > (symSpell.dictionary[$1.lowercased()] ?? 0) }
            for completion in sorted.prefix(limit) {
                let capped = Self.applyCapitalizationTemplate(from: context.currentWord, to: completion)
                if capped.lowercased() != word {
                    results.append(
                        Suggestion(text: capped, score: 0.5, source: .symspell, isUnknownVerbatim: false)
                    )
                }
            }
        } else {
            // Typo correction via SymSpell.
            let corrections = symSpell.lookup(
                input: word,
                verbosity: .top
            )

            for (term, _, distance) in corrections.prefix(limit) {
                let capped = Self.applyCapitalizationTemplate(from: context.currentWord, to: term)
                if capped.lowercased() != word {
                    let score: Double
                    if distance == 0 {
                        score = 1.0
                    } else {
                        score = QwertyGeometry.score(
                            typed: word,
                            candidate: term,
                            symSpellDistance: distance,
                            beta: SharedConfig.Defaults.qwertyDistanceBeta,
                            doublingDiscount: SharedConfig.Defaults.qwertyDoublingDiscount,
                            transpositionDiscount: SharedConfig.Defaults.qwertyTranspositionDiscount
                        )
                    }
                    results.append(
                        Suggestion(text: capped, score: score, source: .symspell, isUnknownVerbatim: false)
                    )
                }
            }
        }

        return results
    }

    deinit {
        FileLogger.shared.info(.prediction, "SymSpellProvider deinit")
    }
}

File: scripts/prediction-sim/lib/symspell-provider.mjs
// Pure-logic JS port of keyboard/Sources/Prediction/SymSpell/SymSpellProvider.swift.
// Kept in sync per AGENTS.md -> Test policy.

import { score as qwertyScore } from './qwerty-geometry.mjs';
import { expansion } from './contractions.mjs';
import { expansion as ambiguousExpansion } from './ambiguous-contractions.mjs';
import { applyCapitalizationTemplate } from './apply-capitalization-template.mjs';

/**
 * Default parameter values matching SharedConfig.Defaults:
 *   qwertyDistanceBeta        = 1.5
 *   qwertyDoublingDiscount    = 0.5
 *   qwertyTranspositionDiscount = 0.7
 */
export const DEFAULTS = {
  beta: 1.5,
  doublingDiscount: 0.5,
  transpositionDiscount: 0.7,
};

/**
 * Adapts SymSpell to the scored-suggestion pipeline.
 *
 * In the Swift, the provider handles:
 *   - Verbatim candidate (always at score 1.0)
 *   - Contraction expansions (score 0.9)
 *   - Prefix completions from Trie for real (known) words (score 0.5)
 *   - QwertyGeometry-scored SymSpell corrections for unknown words
 *
 * In the JS harness, the Trie is not available, so the isRealWord branch
 * returns only the verbatim + contraction entries (no prefix completions).
 * When the word is NOT a real word, SymSpell candidates are scored using
 * QwertyGeometry.
 *
 * @param {import('./symspell.mjs').SymSpell} symSpell - A SymSpell instance
 * @param {Map<string, {count: number}>} [dictionary] - Full dictionary map for
 *   the isRealWord check (word -> {count}). Defaults to symSpell.dictionary.
 * @param {number} [realWordThreshold=2000] - Minimum count for a word to be
 *   considered "real" (known dictionary word).
 */
export class SymSpellProvider {
  constructor(symSpell, dictionary = null, realWordThreshold = 2000) {
    /** @type {import('./symspell.mjs').SymSpell} */
    this._symSpell = symSpell;
    /** @type {Map<string, {count: number}>} */
    this._dictionary = dictionary;
    /** @type {number} */
    this._realWordThreshold = realWordThreshold;
  }

  /**
   * Returns scored suggestions for the given input word.
   *
   * @param {string} word - The word to look up (will be lowercased).
   * @param {object} [opts]
   * @param {number} [opts.limit=8] - Max candidates to return.
   * @param {string} [opts.currentWord=word] - Original casing for the word.
   * @param {'top'|'all'|'closest'} [opts.verbosity='top'] - SymSpell verbosity.
   *   Default 'top' matches the Swift provider (only the best SymSpell candidate).
   *   Pass 'all' for broader candidate exploration in sweeps.
   * @returns {Array<{text: string, score: number, source: string, distance: number, isUnknownVerbatim: boolean}>}
   *   Sorted by score descending.
   */
  suggest(word, opts = {}) {
    const { limit = 8, currentWord = word, verbosity = 'top' } = opts;
    const lower = word.toLowerCase();
    if (!lower) return [];

    // Determine if the typed word is a "real" (known) dictionary word.
    const isRealWord = this._isRealWord(lower);

    // Contraction fast-path: checked BEFORE SymSpell so that
    // apostrophe-less forms like "dont" produce "don't" deterministically.
    const contract = expansion(lower);

    // When a contraction exists, it should be the PRIMARY candidate (leftmost
    // chip, highest score). The verbatim is demoted so it appears as a
    // secondary option the user can tap to keep the apostrophe-less form.
    const verbatimScore = contract ? 0.5 : 1.0;

    /** @type {Array<{text: string, score: number, source: string, distance: number, isUnknownVerbatim: boolean}>} */
    const results = [
      { text: currentWord, score: verbatimScore, source: 'symspell', distance: 0, isUnknownVerbatim: !isRealWord },
    ];

    if (contract) {
      results.unshift({
        text: contract,
        score: 1.0,
        source: 'contraction',
        distance: 0,
        isUnknownVerbatim: false,
      });
    }

    // Ambiguous contraction: the typed token is a real dictionary word with its
    // own meaning ("its" = possessive, "cant" = hypocritical talk), so a
    // deterministic flip would corrupt correct usage. The contraction form is
    // offered as a candidate at 0.5 — injected REGARDLESS of isRealWord,
    // alongside (not instead of) the trie-completion branch — and competes via
    // KenLM fusion. Autocorrect applies it only when the LM-margin gate in
    // topCorrection passes.
    const ambiguous = ambiguousExpansion(lower);
    if (ambiguous) {
      results.push({
        text: applyCapitalizationTemplate(currentWord, ambiguous),
        score: 0.5,
        source: 'ambiguousContraction',
        distance: 0,
        isUnknownVerbatim: false,
      });
    }

    if (isRealWord) {
      // Prefix completions from trie — not available in JS harness.
      // In the Swift, this adds trie completions at score 0.5.
      // Since we don't have a Trie, we just return verbatim + contraction.
      // This matches the current sweep behavior (the Apple-like guard
      // in runOne prevents correction of well-known words).
    } else {
      // Typo correction via SymSpell.
      // Always use 'all' internally so we can find the best non-identical
      // candidate. The input itself may appear at distance 0 (if it's in
      // the dictionary as a low-frequency word), which the identically-
      // named Swift filter `capped.lowercased() != word` excludes. Using
      // 'all' ensures the next-best candidate (distance > 0) is reachable.
      const allCorrections = this._symSpell.lookup(lower, undefined, 'all');

      // Filter out the input itself, then apply verbosity.
      const filtered = allCorrections.filter(c => c.term !== lower);
      const selected = verbosity === 'top'
        ? filtered.slice(0, 1)
        : filtered;

      for (const c of selected) {
        const qScore = qwertyScore(
          lower,
          c.term,
          c.distance,
          DEFAULTS.beta,
          DEFAULTS.doublingDiscount,
          DEFAULTS.transpositionDiscount,
        );
        results.push({
          text: c.term,
          score: qScore,
          source: 'symspell',
          distance: c.distance,
          isUnknownVerbatim: false,
        });
      }
    }

    // Sort by score descending.
    results.sort((a, b) => b.score - a.score);

    // Keep up to limit items, but always include the verbatim entry
    // (it should naturally be near the top with score 1.0).
    return results.slice(0, limit + 1);
  }

  /**
   * Check if a word is a "real" (known) dictionary word.
   * Mirrors the Swift's `trie.contains(word:) || LearnedWordsStore.shared.contains(word)`.
   * In the JS harness, we use dictionary presence with a minimum count threshold.
   *
   * @param {string} lower - Lowercased word.
   * @returns {boolean}
   */
  _isRealWord(lower) {
    if (!this._dictionary) return false;
    const entry = this._dictionary.get(lower);
    if (!entry) return false;
    return entry.count >= this._realWordThreshold;
  }
}
